---
title: 'TE Modeling'
author: 'Scott Teresi'

output:
  pdf_document: default
  word_document:
  html_document:
    toc: yes
    toc_depth: 3
---

# Goals:
Begin playing around with linear models and read into AIC to begin initial modeling exercises.

# Set-Up Module
```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = "~/home/scott/Documents/Uni/Research/transposon_2.0/Camarosa_TE/CAMDATA/") # Path
```



# Install and Load Modules
```{r Install Modules, eval=FALSE, include=FALSE, results='hide'}
    install_modules = function(){
    install.packages("stringi")
    install.packages("tidyr")
    install.packages("dplyr")
    install.packages("reshape2")
    install.packages("cowplot")
    #install_modules()
    }
```

```{r Load Modules, include=FALSE}
  load_modules = function(){
    library("stringi")
    library("tidyr")
    library("dplyr")
    library("reshape2")
    library("cowplot")
  }

  #load_modules()
```

# Load the data
```{r echo=TRUE}
setwd("/home/scott/Documents/Uni/Research/transposon_2.0/Camarosa_TE/CAMDATA/")
directory = setwd("/home/scott/Documents/Uni/Research/transposon_2.0/Camarosa_TE/CAMDATA/") # Path
TE_Data <- list.files(path=directory, pattern="*density_data.csv") # Find the files
TE_Data <- do.call(rbind,lapply(TE_Data,read.csv)) # Bind the data into one set

# Declare the Averages
TE_Data$DNAAVG <- TE_Data$DNA_left+TE_Data$DNA_right 
TE_Data$LTRAVG <- TE_Data$LTR_left+TE_Data$LTR_right 

# Clean up the data
TE_Data <- select(TE_Data, -number) # remove the numbers column, it is vestigial
colnames(TE_Data)[colnames(TE_Data)=="Unknown_fam_left"] <- "UnknownFam_left"
colnames(TE_Data)[colnames(TE_Data)=="Unknown_fam_intra"] <- "UnknownFam_intra"
colnames(TE_Data)[colnames(TE_Data)=="Unknown_fam_right"] <- "UnknownFam_right"

# Take a subset of data for testing purposes, don't want to eat up RAM
#TE_Data = TE_Data %>% slice(1:10000)

# Reformat the data
## One observation per row
#Reshaped<-melt(data = TE_Data,id=c("chromosome","maker_name","start","stop","prox_left","prox_right","they_are_inside","length","window_size"))
#rm(TE_Data) # remove TE_Data because we need the RAM
```
# Get Expression Data
```{r}
setwd("/home/scott/Documents/Uni/Research/transposon_2.0/Camarosa_TE/Expression/test_graphs/") # specify the file path


Expression <- read.csv('expression_out.csv', header=TRUE, sep=",") # read the expression csv
Expression$TPM_AVG <- rowMeans(Expression[c('TPM_0','TPM_1','TPM_2','TPM_3','TPM_4','TPM_5')]) # calculate the mean for TPM
Expression$FPKM_AVG <- rowMeans(Expression[c('FPKM_0','FPKM_1','FPKM_2','FPKM_3','FPKM_4','FPKM_5')]) # calculate the mean for FPKM
GeneExpressionDF<-Expression %>% select(name,TPM_AVG,FPKM_AVG) # make a new data frame with only the TPM and FPKM averages
colnames(GeneExpressionDF)<-c("maker_name","TPM_AVG","FPKM_AVG") # reset the name column to be maker_name
rm(Expression)
```





# Begin Modeling
```{r Model, echo=TRUE}

test_data = randu
test_data$A = 2*test_data$x + test_data$y
lm(test_data$A ~ test_data$x + test_data$y)
#Output comes out as expected.


lm(GeneExpressionDF$TPM_AVG ~ TE_Data$DNA_right)
```












```{r eval=FALSE}
# Other




# Remove the 0s from the data frame
Reshaped <- Reshaped[Reshaped$value!=0,]

#Reshaped$value <- Reshaped$value/2 # stop dividng my two
Reshaped = Reshaped[Reshaped$value<=1,]

#TE.Density.Means<-Reshaped %>% group_by(variable,window_size) %>% summarise(avg=mean(value)/2) %>% arrange(avg)
#Reshaped2 <- Reshaped %>% group_by(variable,window_size) %>% summarise(avg=mean(value)/2) %>% arrange(avg)



# Split TE type/family and location (e.g. left, right, intragenic) in hacky way (reverse string, split by "_", reverse strings again)
#TE.Density.Means$variable <-stri_reverse(TE.Density.Means$variable)
#TE.Density.Means<-TE.Density.Means%>%separate(variable,into = c("Location","TE_type"),sep="_",extra="merge")
#TE.Density.Means$Location<-stri_reverse(TE.Density.Means$Location)
#TE.Density.Means$TE_type<-stri_reverse(TE.Density.Means$TE_type)

# Reorder Location factors so that plot shows left, intra, right
#TE.Density.Means$Location <- factor(TE.Density.Means$Location, levels=c("left", "intra", "right"))

# Make left side windows negative
#TE.Density.Means[TE.Density.Means$Location=="left",]$window_size<-TE.Density.Means[TE.Density.Means$Location=="left",]$window_size*-1

# Rename the middle column to have a capital I
#TE.Density.Means[TE.Density.Means$Location=="intra",]$window_size<-"0"

# Subset to look only at TE types
#TE.Type.Density.Means<-TE.Density.Means[TE.Density.Means$TE_type=="DNA"|TE.Density.Means$TE_type=="LINE"|TE.Density.Means$TE_type=="LTR"|TE.Density.Means$TE_type=="Unknown"|TE.Density.Means$TE_type=="LTRUnknown",]

#Subset to look only at TE families and rename column to TE_family

#TE.Family.Density.Means<-TE.Density.Means[TE.Density.Means$TE_type=="Unknown_fam"|TE.Density.Means$TE_type=="PIF_Harbinger"|TE.Density.Means$TE_type=="LINE_fam"|TE.Density.Means$TE_type=="MULE"|TE.Density.Means$TE_type=="Copia"|TE.Density.Means$TE_type=="Gypsy"|TE.Density.Means$TE_type=="hAT"|TE.Density.Means$TE_type=="CMC_EnSpm",]
#colnames(TE.Family.Density.Means)[2]<-"TE_Family" # LINE fam not none

#============================================================================
setwd("/home/scott/Documents/Uni/Research/transposon_2.0/Camarosa_TE/Expression/test_graphs/") # specify the file path


Expression <- read.csv('expression_out.csv', header=TRUE, sep=",") # read the expression csv
Expression$TPM_AVG <- rowMeans(Expression[c('TPM_0','TPM_1','TPM_2','TPM_3','TPM_4','TPM_5')]) # calculate the mean for TPM
Expression$FPKM_AVG <- rowMeans(Expression[c('FPKM_0','FPKM_1','FPKM_2','FPKM_3','FPKM_4','FPKM_5')]) # calculate the mean for FPKM
GeneExpressionDF<-Expression %>% select(name,TPM_AVG,FPKM_AVG) # make a new data frame with only the TPM and FPKM averages
colnames(GeneExpressionDF)<-c("maker_name","TPM_AVG","FPKM_AVG") # reset the name column to be maker_name


#setwd("/home/scott/Documents/Uni/Research/transposon_2.0/Camarosa_TE/Expression/test_graphs/") # specify the file path

# Take a subset of Reshaped by a variable
TE.TypeExp <-Reshaped %>% filter(variable == "LTR_left")
TE.TypeExp <- TE.TypeExp %>%  left_join(GeneExpressionDF, by="maker_name")
TE.TypeExp <- TE.TypeExp[TE.TypeExp$value<=1,] # filter out things with a density above 1
pdf('LTR_left.pdf')
ggplot(TE.TypeExp,aes(x=value,y=log2(TPM_AVG),color=as.character(window_size)))+geom_smooth(method='auto')+facet_wrap(~window_size)+
labs(title="LTR Tranposon Density Upstream", x = "Tranposon Density by Window", y= "Gene Expression (Log(2) TPM)", color="Window Size")+
scale_x_continuous(breaks=seq(0,1))
dev.off()

```